lib/teamsnap.rb

```diff
@@ -131,18 +136,50 @@ def init(opts = {})
       apply_endpoints(self, collection) && true
     end
 
-    def run(via, href, args = {})
+    def run(via, href, args = {}, opts = {})
       resp = client.send(via, href, args)
 
       if resp.status == 200
-        Oj.load(resp.body)
-          .fetch(:collection)
+        collection = Oj.load(resp.body).fetch(:collection)
+        TeamSnap.write_backup_file(opts[:backup_cache_file], collection)
+        collection
       else
-        error_message = parse_error(resp)
-        raise TeamSnap::Error.new(error_message)
+        if TeamSnap.backup_file_exists?(opts[:backup_cache_file])
+          warn("Connection to API failed.. using backup cache file to initialize endpoints")
+          Oj.load(IO.read(opts[:backup_cache_file]))
+        else
+          error_message = parse_error(resp)
+          raise TeamSnap::Error.new(error_message)
+        end
+      end
+    end
+
+    def backup_file_for(backup_cache)
+      case backup_cache
+      when true
+        "./tmp/.teamsnap_rb"
+      else
+        backup_cache
+      end
+    end
+
+    def write_backup_file(file_location, collection)
+      return unless file_location
```

Comment:
This is related to the previous comment of "tell, don't ask" - I was checking if the backup_cache_file was set (aka, false was not passed) prior to calling the write method. Since backup_cache_file is handled internally, there is no error needed (the only time this is called without a file_location is when we do not want to save the backup file)